### DISCUSSION TITLE
x-for with nested x-data does not refresh properly based on index

### DISCUSSION THREAD

**Discussion by atomgiant** [alpinejs/alpine#1804] [7. Bugs (archived)]
I posted this in Discord originally but thought I'd share this here to get a conversation going.

I ran into an issue porting some code to Alpine v3. The code in question had an x-for with a nested x-data scope inside of it. Alpine is not refreshing the nested x-data when items are removed from the list. You can see this in action here:

 https://codepen.io/atomgiant/pen/vYmxQNK

When an item is removed from the list, such as removing the item at index 0, the other items shift and end up using the same index. This causes the x-data to not change since the index is the same, but the underlying data is not the same. This seems like a bug to me.

In v2 I was able to pass in the item as the key to prevent this, but V3 requires a string or integer for the x-for key.

I was able to work around this by generating an index on the fly for the items in the list. For example, using this for the x-for key:

`<template x-for="(item, index) in items" :key="itemKey(item)">`

Then inside of my data I define these:

```javascript
  itemKeys: new WeakMap(),
  nextItemKey: 0,
  itemKey(item) {
    if (!this.itemKeys.has(item)) {
      this.itemKeys.set(item, this.nextItemKey);
      this.nextItemKey += 1;
    }
    return this.itemKeys.get(item);
  }
```

Is there a better way to handle this within Alpine?

**Comment by KevinBatdorf** [comment]
>  the other items shift and end up using the same index

This is because you're using the `index` as the key. You should _never_ use the index as the key (fyi this goes for React, Vue, etc) but instead use a unique identifier. The purpose of the key is for this exact scenario.

There is actually never a good reason to use the index as the key

**Reply by atomgiant** [reply]
Thanks Kevin. I appreciate why this is happening and I'm not suggesting Alpine should work with just an index in this case.

I'm mainly wondering if there is easier way to handle an `x-for` loop over a list of objects when the objects do not have a unique id.  My workaround to generate an incrementing key based on the Object reference seems to be working well. 

If you know of a better way to handle this I'd appreciate hearing your thoughts.

**Reply by KevinBatdorf** [reply]
From the user's perspective you should just generate a random ID to use. Not at runtime but when the object is created. It could be random-ish even and work fine.The ID just can't change between renders.

```js
{
  name: 'One',
  id: Date.now() + Math.floor(Math.random() * 1000000),
}
```
That would probably be fine.

In your codepen, the key could also just be `item.name`

`:key="itemKey(item)"` - This may also work but in your example code you're not returning the key

**Reply by atomgiant** [reply]
Thanks @KevinBatdorf. I updated my example to add the return. That was accidentally left off when I shared my workaround before.

In my case I have data that doesn't need an ID for anything other than the x-for loop so I'm going to keep using my workaround. Perhaps this thread will be useful for others with a similar use case.
